presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchingVisitor.java

```diff
@@ -38,43 +42,108 @@
-        return super.visitProject(node, context);
+        boolean result = super.visitProject(node, context);
+        if (result) {
+            context.getExpressionAliases().replaceAssignments(node.getAssignments());
+        }
+        return result;
     }
 
     @Override
     protected Boolean visitPlan(PlanNode node, PlanMatchingContext context)
     {
-        List<PlanMatchingState> states = context.getPattern().matches(node, session, metadata, context.getExpressionAliases());
+        List<PlanMatchingState> states = context.getPattern().downMatches(node, context.getExpressionAliases());
 
+        // No shape match; don't need to check the internals of any of the nodes.
         if (states.isEmpty()) {
             return false;
         }
 
+        // Leaf node in the plan.
         if (node.getSources().isEmpty()) {
-            return !filterTerminated(states).isEmpty();
+            int terminatedUpMatchCount = 0;
+            for (PlanMatchingState state : states) {
+                // Don't consider un-terminated PlanMatchingStates.
+                if (!state.isTerminated()) {
+                    continue;
+                }
+
+                /*
+                 * We have to call upMatches for two reasons:
+                 * 1) Make sure there aren't any mismatches checking the internals of a leaf node.
+                 * 2) Calling upMatches has the side-effect of adding aliases to the context's
+                 *    ExpressionAliases. They'll be needed further up.
+                 */
+                if (context.getPattern().upMatches(node, session, metadata, context.getExpressionAliases())) {
+                    ++terminatedUpMatchCount;
+                }
+            }
+
+            checkState(terminatedUpMatchCount < 2, format("Ambiguous shape match on leaf node %s", node));
+            return terminatedUpMatchCount == 1;
         }
 
+        int upMatchCount = 0;
         for (PlanMatchingState state : states) {
             checkState(node.getSources().size() == state.getPatterns().size(), "Matchers count does not match count of sources");
             int i = 0;
             boolean sourcesMatch = true;
+
+            /*
+             * For every state, start with a clean set of aliases. Aliases from a different state
+             * are not in scope.
+             */
+            ExpressionAliases stateAliases = new ExpressionAliases();
             for (PlanNode source : node.getSources()) {
-                sourcesMatch = sourcesMatch && source.accept(this, state.createContext(i++));
+                /*
+                 * Create a context for each source individually. Aliases from one source
+                 * shouldn't be visible in the context of other sources.
+                 */
+                PlanMatchingContext sourceContext = state.createContext(i++);
+                sourcesMatch = sourcesMatch && source.accept(this, sourceContext);
```

Comment:
`sourcesMatch &` is not necessary. `sourceMatch` is always true in this look (since it breaks as soon as it becomes false). I'd rewrite it as:

```java
if (!source.accept(this, sourceContext)) {
   sourcesMatch = false;
   break;
}
```